File: drills/ts/solutions/nc123_alien_dictionary.ts
export function alienOrder(words: string[]): string {
  const letters = new Set<string>();
  for (const word of words) for (const letter of word) letters.add(letter);

  const waiting = new Map<string, number>();
  const unlocks = new Map<string, string[]>();
  for (const letter of letters) {
    waiting.set(letter, 0);
    unlocks.set(letter, []);
  }

  // Two adjacent words agree up to their first difference, and that difference
  // is the only thing they say about the alphabet -- everything after it is
  // unordered. The one case with no letters to compare is a word followed by a
  // prefix of itself, which no alphabet can explain.
  for (let i = 0; i + 1 < words.length; i++) {
    const first = words[i];
    const second = words[i + 1];
    const shared = Math.min(first.length, second.length);
    let split = -1;
    for (let j = 0; j < shared; j++) {
      if (first[j] !== second[j]) {
        split = j;
        break;
      }
    }
    if (split === -1) {
      if (first.length > second.length) return "";
      continue;
    }
    unlocks.get(first[split])!.push(second[split]);
    waiting.set(second[split], waiting.get(second[split])! + 1);
  }

  const ready = [...letters].filter((letter) => waiting.get(letter) === 0);
  const order: string[] = [];
  while (ready.length) {
    const letter = ready.pop()!;
    order.push(letter);
    for (const following of unlocks.get(letter)!) {
      waiting.set(following, waiting.get(following)! - 1);
      if (waiting.get(following) === 0) ready.push(following);
    }
  }

  // Short means the leftovers all depend on each other: the ordering the words
  // describe is contradictory, so no alphabet satisfies it.
  return order.length === letters.size ? order.join("") : "";
}

File: drills/go/solutions/nc123_alien_dictionary.go
package main

func alienOrder(words []string) string {
	// Adjacent words give one ordering each: the first differing letter.
	// Then a topological sort (Kahn) over the letters; a leftover means a cycle.
	next := map[byte]map[byte]bool{}
	indegree := map[byte]int{}
	for _, w := range words {
		for i := 0; i < len(w); i++ {
			if next[w[i]] == nil {
				next[w[i]] = map[byte]bool{}
				indegree[w[i]] = 0
			}
		}
	}
	for i := 0; i+1 < len(words); i++ {
		a, b := words[i], words[i+1]
		if len(a) > len(b) && a[:len(b)] == b {
			return ""
		}
		for j := 0; j < len(a) && j < len(b); j++ {
			if a[j] != b[j] {
				if !next[a[j]][b[j]] {
					next[a[j]][b[j]] = true
					indegree[b[j]]++
				}
				break
			}
		}
	}
	queue := []byte{}
	for c := byte('a'); c <= 'z'; c++ {
		if n, present := indegree[c]; present && n == 0 {
			queue = append(queue, c)
		}
	}
	order := []byte{}
	for len(queue) > 0 {
		c := queue[0]
		queue = queue[1:]
		order = append(order, c)
		for d := byte('a'); d <= 'z'; d++ {
			if next[c][d] {
				indegree[d]--
				if indegree[d] == 0 {
					queue = append(queue, d)
				}
			}
		}
	}
	if len(order) != len(indegree) {
		return ""
	}
	return string(order)
}
